### PR TITLE
Gg 7344

### PR DESCRIPTION
**Pull request by jinbal-hmrc**

### CODE REVIEW

**Review by martin-armstrong — DISMISSED** [review]

**Comment by ModisR** [comment]
This change could have been done more simply by not introducing a new form model. Details follow.

File: app/controllers/DeleteEntryController.scala [hunk 1]
```diff
@@ -17,20 +17,22 @@
...
+import scala.concurrent.{ExecutionContext, Future}
 
 @Singleton
 class DeleteEntryController @Inject() (
   siProtectedUserListService: SiProtectedUserListService,
   views: Views,
   mcc: MessagesControllerComponents,
-  val strideAction: StrideAction
+  val strideAction: StrideAction,
+  inputForms: InputForms
```
Comment (ModisR) [inline app/controllers/DeleteEntryController.scala:35]: This change could be undone.

File: app/controllers/DeleteEntryController.scala [hunk 2]
```diff
@@ -17,20 +17,22 @@
 package controllers
 
 import controllers.base.{StrideAction, StrideController}
+import models.InputForms
```
Comment (ModisR) [inline app/controllers/DeleteEntryController.scala:20]: This change could be undone.

File: app/controllers/DeleteEntryController.scala [hunk 3]
```diff
@@ -43,13 +45,22 @@ class DeleteEntryController @Inject() (
...
+        entryId => {
+          siProtectedUserListService
+            .deleteEntry(entryId)
+            .map(_ => Ok(views.deleteSuccess()))
+            .recover {
+              case UpstreamErrorResponse(_, 404, _, _) => NotFound(views.errorTemplate("delete.entry.not.found", "delete.entry.not.found", "delete.entry.already.deleted"))
+              case _                                   => InternalServerError(views.somethingWentWrong())
+            }
+        }
+      )
```
Comment (ModisR) [inline app/controllers/DeleteEntryController.scala:64]: This change could be undone.

File: app/models/InputForms.scala [hunk 4]
```diff
@@ -58,6 +58,10 @@ class InputForms @Inject() (config: SiProtectedUserConfig) {
     )(Search.apply)(Search.unapply)
   )
 
+  def deleteForm: Form[String] = Form(
+    single("entryId" -> nonEmptyText)
+  )
+
```
Comment (ModisR) [inline app/models/InputForms.scala:64]: This change could be undone.

File: app/views/delete_entry_confirmation.scala.html [hunk 5]
```diff
@@ -24,18 +24,22 @@
...
 
 @(protectedUserRecord: ProtectedUserRecord)(implicit request: Request[_], messages: Messages)
+    @mainLayout(
+        pageTitle = Messages("confirm.delete.entry.title"),
+    ) {
+        @h1("confirm.delete.entry.header")
+        @details(protectedUserRecord)
+        @govukWarningText(WarningText(content = Text(Messages("confirm.delete.warning"))))
+        @formTemplate(action = controllers.routes.DeleteEntryController.delete(), args = Symbol("autocomplete") -> "off", Symbol("novalidate") -> "") {
+            @input(Input(id = "entryId", name = "entryId", inputType = "hidden", value = Some(protectedUserRecord.entryId)))
```
Comment (ModisR) [inline app/views/delete_entry_confirmation.scala.html:40]: Delete form doesn't need an `<input>` element. You can continue passing that through the URL as before this change.

File: conf/app.routes [hunk 6]
```diff
@@ -11,7 +11,7 @@ GET         /edit/:entryId                  controllers.EditEntryController.show
 POST        /edit                           controllers.EditEntryController.submit()
 GET         /view-entry/:entryId            controllers.SiProtectedUserController.view(entryId)
 GET         /confirm-delete/:entryId        controllers.DeleteEntryController.showConfirmDeletePage(entryId)
-GET         /delete-entry/:entryId          controllers.DeleteEntryController.delete(entryId)
+POST        /delete-entry                   controllers.DeleteEntryController.delete()
```
Comment (ModisR) [inline conf/app.routes:14]: The only change needed here is to substitute `GET` with `POST`.

File: it/controllers/DeleteEntryControllerISpec.scala [hunk 7]
```diff
@@ -46,10 +46,10 @@ class DeleteEntryControllerISpec extends BaseISpec with ResultExtractors {
         expectDeleteEntryToBeSuccessful(record)
 
         val response = wsClient
-          .url(resource(s"$frontEndBaseUrl/delete-entry/${record.entryId}"))
+          .url(resource(s"$frontEndBaseUrl/delete-entry"))
```
Comment (ModisR) [inline it/controllers/DeleteEntryControllerISpec.scala:49]: This can be undone.

File: it/controllers/DeleteEntryControllerISpec.scala [hunk 8]
```diff
@@ -46,10 +46,10 @@ class DeleteEntryControllerISpec extends BaseISpec with ResultExtractors {
...
           .withHttpHeaders("Csrf-Token" -> "nocheck")
           .withCookies(mockSessionCookie)
-          .get()
+          .post(Map("entryId" -> record.entryId))
```
Comment (ModisR) [inline it/controllers/DeleteEntryControllerISpec.scala:52]: A `Map.empty` can be posted here.

File: it/controllers/DeleteEntryControllerISpec.scala [hunk 9]
```diff
@@ -66,10 +66,25 @@ class DeleteEntryControllerISpec extends BaseISpec with ResultExtractors {
         )
 
         val response = wsClient
-          .url(resource(s"$frontEndBaseUrl/delete-entry/${record.entryId}"))
+          .url(resource(s"$frontEndBaseUrl/delete-entry"))
```
Comment (ModisR) [inline it/controllers/DeleteEntryControllerISpec.scala:69]: This change can be undone.

File: it/controllers/DeleteEntryControllerISpec.scala [hunk 10]
```diff
@@ -66,10 +66,25 @@ class DeleteEntryControllerISpec extends BaseISpec with ResultExtractors {
...
+          .withHttpHeaders("Csrf-Token" -> "nocheck")
+          .withCookies(mockSessionCookie)
+          .post(Map("entryId" -> record.entryId))
```
Comment (ModisR) [inline it/controllers/DeleteEntryControllerISpec.scala:72]: A `Map.empty` can be posted here.

File: it/controllers/DeleteEntryControllerISpec.scala [hunk 11]
```diff
@@ -66,10 +66,25 @@ class DeleteEntryControllerISpec extends BaseISpec with ResultExtractors {
...
+      forAll(protectedUserRecords, nonEmptyStringGen) { (record, pid) =>
+        expectUserToBeStrideAuthenticated(pid)
+
+        val response = wsClient
+          .url(resource(s"$frontEndBaseUrl/delete-entry"))
           .withHttpHeaders("Csrf-Token" -> "nocheck")
           .withCookies(mockSessionCookie)
-          .delete()
+          .post(Map.empty[String, String])
           .futureValue
```
Comment (ModisR) [inline it/controllers/DeleteEntryControllerISpec.scala:88]: This test can be deleted.

File: test/controllers/DeleteEntryControllerSpec.scala [hunk 12]
```diff
@@ -29,7 +29,8 @@ class DeleteEntryControllerSpec extends BaseControllerSpec {
       mockBackendService,
       injectViews,
       Stubs.stubMessagesControllerComponents(),
-      stubStrideActions.sample.get
+      stubStrideActions.sample.get,
+      inputForms
```
Comment (ModisR) [inline test/controllers/DeleteEntryControllerSpec.scala:33]: This change can be undone.

File: test/controllers/DeleteEntryControllerSpec.scala [hunk 13]
```diff
@@ -66,7 +67,7 @@ class DeleteEntryControllerSpec extends BaseControllerSpec {
             mockBackendService.deleteEntry(eqTo(record.entryId))(*, *)
           } thenReturn Future.successful(record)
 
-          val result = deleteEntryController.delete(record.entryId)(FakeRequest().withMethod("DELETE"))
+          val result = deleteEntryController.delete()(FakeRequest().withFormUrlEncodedBody("entryId" -> record.entryId).withMethod("POST"))
```
Comment (ModisR) [inline test/controllers/DeleteEntryControllerSpec.scala:70]: The only change needed here is to substitute "DELETE" with "POST".

File: test/controllers/DeleteEntryControllerSpec.scala [hunk 14]
```diff
@@ -84,7 +85,7 @@ class DeleteEntryControllerSpec extends BaseControllerSpec {
             mockBackendService.deleteEntry(eqTo(record.entryId))(*, *)
           } thenReturn Future.failed(UpstreamErrorResponse("not found", NOT_FOUND))
 
-          val result = deleteEntryController.delete(record.entryId)(FakeRequest().withMethod("DELETE"))
+          val result = deleteEntryController.delete()(FakeRequest().withFormUrlEncodedBody("entryId" -> record.entryId).withMethod("POST"))
```
Comment (ModisR) [inline test/controllers/DeleteEntryControllerSpec.scala:88]: The only change needed here is to substitute "DELETE" with "POST".

File: test/controllers/DeleteEntryControllerSpec.scala [hunk 15]
```diff
@@ -94,14 +95,28 @@ class DeleteEntryControllerSpec extends BaseControllerSpec {
...
+
+          val result = deleteEntryController.delete()(FakeRequest().withMethod("POST"))
+
+          status(result) shouldBe NOT_FOUND
+          val body = contentAsString(result)
+          body should include("delete.entry.not.found")
+          body should include("delete.entry.not.found")
+        }
+      }
+    }
```
Comment (ModisR) [inline test/controllers/DeleteEntryControllerSpec.scala:110]: This scenario doesn't need a test.

File: test/controllers/DeleteEntryControllerSpec.scala [hunk 16]
```diff
@@ -94,14 +95,28 @@ class DeleteEntryControllerSpec extends BaseControllerSpec {
...
+
     "Forward to error page when delete is unsuccessful" in {
       forAll(protectedUserRecords) { record =>
         expectStrideAuthenticated {
           when {
             mockBackendService.deleteEntry(eqTo(record.entryId))(*, *)
           } thenReturn Future.failed(UpstreamErrorResponse("internal server error", INTERNAL_SERVER_ERROR))
 
-          val result = deleteEntryController.delete(record.entryId)(FakeRequest().withMethod("DELETE"))
+          val result = deleteEntryController.delete()(FakeRequest().withFormUrlEncodedBody("entryId" -> record.entryId).withMethod("POST"))
```
Comment (ModisR) [inline test/controllers/DeleteEntryControllerSpec.scala:119]: The only change needed here is to substitute "DELETE" with "POST".

**Review by ModisR — CHANGES_REQUESTED** [review]

**Review by ModisR — APPROVED** [review]

**Pull request merged**